### PR TITLE
[.Net] Add Goolge gemini

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
## Related issue number
#2851 task 1,2,3,4,5

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by LittleLittleCloud** [comment]
@hsm207 could you please review the examples under `sample/AutoGen.Gemini.Sample`

**Comment by hsm207** [comment]
@LittleLittleCloud I ran all the Vertex-related samples. Here is what I found:

1. All samples run as expected except `Image_Chat_With_Vertex_Gemini`
2. For `Image_Chat_With_Vertex_Gemini`, the path to the image is wrong, it should be:

```

var imagePath = Path.Combine("..", "..", "resource", "images", "background.png");
```

3. For `Image_Chat_With_Vertex_Gemini`, both images in the `/resource/images` folder give the following error:

```
Unhandled exception. System.AggregateException: One or more errors occurred. (Status(StatusCode="InvalidArgument", Detail="Provided image is not valid."))
 ---> Grpc.Core.RpcException: Status(StatusCode="InvalidArgument", Detail="Provided image is not valid.")
   at Grpc.Net.Client.Internal.HttpContentClientStreamReader`2.MoveNextCore(CancellationToken cancellationToken)
   at Google.Api.Gax.Grpc.AsyncResponseStream`1.MoveNextAsync(CancellationToken cancellationToken)
   at AutoGen.Gemini.GeminiChatAgent.GenerateStreamingReplyAsync(IEnumerable`1 messages, GenerateReplyOptions options, CancellationToken cancellationToken)+MoveNext() in /workspaces/autogen/dotnet/src/AutoGen.Gemini/GeminiChatAgent.cs:line 151
   at AutoGen.Gemini.GeminiChatAgent.GenerateStreamingReplyAsync(IEnumerable`1 messages, GenerateReplyOptions options, CancellationToken cancellationToken)+MoveNext() in /workspaces/autogen/dotnet/src/AutoGen.Gemini/GeminiChatAgent.cs:line 151
   at AutoGen.Gemini.GeminiChatAgent.GenerateStreamingReplyAsync(IEnumerable`1 messages, GenerateReplyOptions options, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at AutoGen.Gemini.Middleware.GeminiMessageConnector.InvokeAsync(MiddlewareContext context, IStreamingAgent agent, CancellationToken cancellationToken)+MoveNext() in /workspaces/autogen/dotnet/src/AutoGen.Gemini/Middleware/GeminiMessageConnector.cs:line 48
   at AutoGen.Gemini.Middleware.GeminiMessageConnector.InvokeAsync(MiddlewareContext context, IStreamingAgent agent, CancellationToken cancellationToken)+MoveNext() in /workspaces/autogen/dotnet/src/AutoGen.Gemini/Middleware/GeminiMessageConnector.cs:line 48
   at AutoGen.Gemini.Middleware.GeminiMessageConnector.InvokeAsync(MiddlewareContext context, IStreamingAgent agent, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at AutoGen.Core.PrintMessageMiddleware.InvokeAsync(MiddlewareContext context, IStreamingAgent agent, CancellationToken cancellationToken)+MoveNext() in /workspaces/autogen/dotnet/src/AutoGen.Core/Middleware/PrintMessageMiddleware.cs:line 54
   at AutoGen.Core.PrintMessageMiddleware.InvokeAsync(MiddlewareContext context, IStreamingAgent agent, CancellationToken cancellationToken)+MoveNext() in /workspaces/autogen/dotnet/src/AutoGen.Core/Middleware/PrintMessageMiddleware.cs:line 54
   at AutoGen.Core.PrintMessageMiddleware.InvokeAsync(MiddlewareContext context, IStreamingAgent agent, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at AutoGen.Core.PrintMessageMiddleware.InvokeAsync(MiddlewareContext context, IAgent agent, CancellationToken cancellationToken) in /workspaces/autogen/dotnet/src/AutoGen.Core/Middleware/PrintMessageMiddleware.cs:line 24
   at AutoGen.Core.PrintMessageMiddleware.InvokeAsync(MiddlewareContext context, IAgent agent, CancellationToken cancellationToken) in /workspaces/autogen/dotnet/src/AutoGen.Core/Middleware/PrintMessageMiddleware.cs:line 24
   at AutoGen.Core.AgentExtension.SendAsync(IAgent agent, IMessage message, IEnumerable`1 chatHistory, CancellationToken ct) in /workspaces/autogen/dotnet/src/AutoGen.Core/Extension/AgentExtension.cs:line 39
   at AutoGen.Core.AgentExtension.SendAsync(IAgent agent, String message, IEnumerable`1 chatHistory, CancellationToken ct) in /workspaces/autogen/dotnet/src/AutoGen.Core/Extension/AgentExtension.cs:line 59
   at AutoGen.Gemini.Sample.Image_Chat_With_Vertex_Gemini.RunAsync() in /workspaces/autogen/dotnet/sample/AutoGen.Gemini.Sample/Image_Chat_With_Vertex_Gemini.cs:line 37
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Program.<Main>$(String[] args) in /workspaces/autogen/dotnet/sample/AutoGen.Gemini.Sample/Program.cs:line 9
```
I tried with a different image and in jpg format, and it works.

Thanks for making this PR so quickly!

**Review by hsm207 — COMMENTED** [review]
see comment in PR

**Comment by LittleLittleCloud** [comment]
@hsm207 Thanks for the feedback, I'll take a look

> For Image_Chat_With_Vertex_Gemini, the path to the image is wrong

Looks like you are running this example on a unix system, I wonder this error might be because the resource path is configured in a windows-path-like way (forward `\`) in Directory.Build.Props, I'll take a look

**Comment by hsm207** [comment]
@LittleLittleCloud I was running it inside the full devcontainer.

**Comment by LittleLittleCloud** [comment]
> I was running it inside the full devcontainer. I think I got that error because I ran the examples from the `AutoGen.Gemini.Sample` folder and am guessing you did it from the `dotnet` folder?

Oh you are right, that's probably right. Because if you run the project under sample folder, then the `Directory.Build.Props` won't taken into effect

> I tried with a different image and in jpg format, and it works.

It's probably because of `git lfs` so the image in repo is actually a hash code instead of an image?

**Comment by hsm207** [comment]
> It's probably because of `git lfs` so the image in repo is actually a hash code instead of an image?

You're right! I did not notice it was just a hash code before. After pulling all the lfs objects, the sample with those image works.

**Review by DavidLuong98 — APPROVED** [review]

File: dotnet/src/AutoGen.Gemini/Extension/FunctionContractExtension.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                            parameterProperties,
+                        },
+                Type = OpenAPISchemaType.Object,
+            },
+        };
+    }
+
+    private static OpenApiSchema ToOpenApiSchema(Type type)
+    {
+        if (type == null)
```
[the remark this answers is not in the document]
Comment (LittleLittleCloud) [reply dotnet/src/AutoGen.Gemini/Extension/FunctionContractExtension.cs:57]: Sure

**Pull request merged**